src/python/WMCore/Lexicon.py

```diff
@@ -710,6 +711,103 @@ def activity(candidate):
     raise AssertionError("Invalid dashboard activity: %s should 'test'" % candidate)
 
 
+def gpuParameters(candidate):
+    """
+    Validate the spec "GPUParams" argument, which is a JSON encoded object, thus:
+    * an encoded None object (like 'null')
+    * an encoded dictionary with the following parameters:
+      * mandatory: GPUMemoryMB (int), CUDARuntime (str), CUDACapabilities (list of str)
+      * optional: GPUName (str), CUDADriverVersion (str), CUDARuntimeVersion (str)
+    :param candidate: a JSON encoded data to be validated
+    :return: True if validation succeeded, False or exception otherwise
+    """
+    mandatoryArgs = set(["GPUMemoryMB", "CUDARuntime", "CUDACapabilities"])
+    optionalArgs = set(["GPUName", "CUDADriverVersion", "CUDARuntimeVersion"])
+    try:
+        data = json.loads(candidate)
+    except Exception:
+        raise AssertionError("GPUParams is not a valid JSON object")
+        # once python2 code is deprecated, this is the way to raise only the last exception
+        # raise AssertionError("GPUParams is not a valid JSON object") from None
+    if data is None:
+        return True
+    if not isinstance(data, dict):
+        raise AssertionError("GPUParams is not a dictionary encoded as JSON object")
+
+    paramSet = set(viewkeys(data))
+    # is every mandatory argument also in the provided args?
+    if not mandatoryArgs <= paramSet:
+        msg = "GPUParams does not contain all the mandatory arguments. "
+        msg +="Mandatory args: {}, while args provided are: {}".format(mandatoryArgs, paramSet)
+        raise AssertionError(msg)
+    # are there unknown arguments in the data provided?
+    unknownArgs = paramSet - mandatoryArgs - optionalArgs
+    if unknownArgs:
+        msg = "GPUParams contains arguments that are not supported. Args provided: {}, ".format(paramSet)
+        msg +="while mandatory args are: {} and optional args are: {}".format(mandatoryArgs, optionalArgs)
+        raise AssertionError(msg)
+    return _gpuInternalParameters(data)
+
+
+CUDA_VERSION_REGEX = {"re": r"^\d+\.\d+(\.\d+)?$", "maxLength": 100}
```

Comment:
if this one  is about to be in the global scope, I'd say we should move it at the top of the file similar to others defined on lines 27-31

Reply:
That could be done as well. However, looking at this module, you can see that the regular expression is usually defined closer to the function that will consume it, so I just kept the consistency.
Anyhow, this will be refactored once a decision is made on how to separate Lexicon logic from lexicon rules/regex.